Added feature to allow building or downloading libsodium.

Fraser999 opened the pull request:
This adds a feature `get-libsodium` which will either download the prebuilt binaries if running MinGW on Windows, or clone libsodium and build from source if not on Windows (see #131).

Currently MSVC builds aren't handled for this feature, but if there is an interest in using this PR, we can probably add support for MSVC builds too if desired.

jedisct1 commented:
Very nice.

mcginty commented on libsodium-sys/libsodium:
```diff
@@ -0,0 +1 @@
+Subproject commit f01d7c128214e8f976889c8093a42d9cae1abe4b
```
would it make more sense to "lock" the commit sodiumoxide chooses to link against to tagged releases? https://github.com/jedisct1/libsodium/releases (https://github.com/jedisct1/libsodium/commit/fce6852d64339efa33c0ee4130b3107b888d6067 is the latest, 1.0.10)

Fraser999 replied:
Yes, I was in two minds about this.

It would be good to have Windows and non-Windows builds using the exact same build of libsodium, and using the tagged commit would achieve that.

However, I was really keen to use the threadsafe version of `sodium_init()` which will be included in 1.0.11 since this still isn't handled by sodiumoxide, so I selfishly just used tip-of-tree of the stable branch of libsodium.

Ultimately I guess this will be a decision for @dnaq since he'll be the one updating the submodule commit and the tag version for Windows builds on a regular basis if this PR is accepted.  If the general preference is to use the tagged commit, we could probably reduce the maintenance overhead by defining the tag version outside of the platform-specific functions, then for non-Windows builds, do something like a `git fetch` then `git checkout $VERSION`.  That way, only the `VERSION` variable would need to be updated as and when a new release of libsodium lands.

mcginty replied:
Ahh, I didn't know about what was coming in 1.0.11. Nifty!

I'd say from the conservative perspective, using specific stable releases whenever possible in cryptography libraries makes me do less of a double-take though (and it looks like the author signs releases, but not individual commits). I suppose this will be a non-issue once 1.0.11 is released officially.

dnaq replied:
I'll just hold off on the merge until 1.0.11 is released. Thanks for all
your work.
On Thu, 28 Jul 2016 at 11:48, Jake McGinty <notifications@github.com> wrote:

> In libsodium-sys/libsodium
> <https://github.com/dnaq/sodiumoxide/pull/135#discussion_r72593772>:
>
> > @@ -0,0 +1 @@
> > +Subproject commit f01d7c128214e8f976889c8093a42d9cae1abe4b
>
> Ahh, I didn't know about what was coming in 1.0.11. Nifty!
>
> I'd say from the conservative perspective, using specific stable releases
> whenever possible in cryptography libraries makes me do less of a
> double-take though (and it looks like the author signs releases, but not
> individual commits). I suppose this will be a non-issue once 1.0.11 is
> released officially.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dnaq/sodiumoxide/pull/135/files/cf42804a26463f8adb595529a33d1491c77e3a50#r72593772>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPVBT-f-sBB-xtaTnuG8G30HSl4mQruks5qaHrSgaJpZM4JJZI9>
> .
>

mcginty replied:
Looks like 1.0.11 is out now

Fraser999 commented:
I've updated this a bit in the following ways:

1. Both Windows and non-Windows builds use the [libsodium release tarballs](https://download.libsodium.org/libsodium/releases/).  This avoids maintaining a git submodule - updating the libsodium version should be a single line change to the build script for all build types.  I did look at using the tagged commit as @mcginty suggested above, but the tagged commit hasn't had autoconf run on it - i.e. it's not straightforward to build from this commit.  It looks like the 'stable' branch is really the one intended for users building from source.
2. I changed how powershell downloads the tarball - the `wget` command I was using doesn't exist in the default powershell version on Windows 7.
3. On Windows, the doc tests fail to build if pthread is not explicitly linked (I'm not sure why building the doc tests is different to building the lib tests).  I added a section to the build script which guesses at the location of the MinGW pthread lib on the basis of the path to gcc.  This feels a bit flaky to me - it should maybe be a best-effort attempt, allowing sodiumoxide to still be built even though the doc tests would fail.

kpcyrd commented:
You've put a lot of effort into this, and I would suggest to have a separate package that supports something like this for simplicity.

I have some concerns with this, as it downloads binaries from an external server without signature verification (besides https, which doesn't help if the server is compromised or if somebody is able to generate valid certificates). Also, even if signature verification is added, some users might run into this accidentally and would prefer to have it fail and then install the necessary library from the repository of their distribution, instead of pulling executable binaries from the internet.

mcginty commented:
@kpcyrd It's not downloading binaries, it's downloading source and building (maybe you're thinking of the homebrew model ;)

As far as validating the code, how can this not be solved while keeping it in-package by simply "pinning" the libsodium signing key in this repo and then downloading the corresponding .sig file and verifying it's valid and from the known key? It can be an optional step and just print a warning that it couldn't validate if gpg isn't installed.

Also, curl has a `--pinnedpubkey` flag to be a little more rigid about what it's willing to download.

dnaq commented:
I know it's pretty late, but I've been looking at merging this pull request. However, it would be much nicer if we didn't have to hardcode the verison number of libsodium, and also if we could somehow verify that we have gotten the right tar-file before unpacking and installing.

I know that you maintain your own fork of sodiumoxide nowadays, so I don't know how interested you would be in updating this. If you're not, let me know so I can close the issue.

ffflorian commented:
So there is no chance this will get merged any time soon, right? :slightly_frowning_face:

dnaq commented:
Sorry, no it won't.
